### DISCUSSION TITLE
Multicloud K3s not working

### DISCUSSION THREAD

**Discussion by AlbertoSoutullo** [k3s-io/k3s#8657] [Q&A] [answered]
**Environmental Info:**
K3s Version: 
k3s version v1.27.6+k3s1 (bd04941a)
go version go1.20.8

Node(s) CPU architecture, OS, and Version: 
Server: Linux  5.15.0-75-generic #82-Ubuntu SMP Tue Jun 6 23:10:23 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
Agent: Linux 5.15.0-84-generic #93-Ubuntu SMP Tue Sep 5 17:16:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:
2 Nodes in different networks.

**Describe the bug:**
Hello, first of all, my experience with Kubernetes is not very extensive. I am having troubles setting up a [[multicloud](https://docs.k3s.io/installation/network-options#embedded-k3s-multicloud-solution)](https://docs.k3s.io/installation/network-options#embedded-k3s-multicloud-solution) cluster with K3s. We have several baremetal machines in different networks. I want those machines to work in the same cluster. We have wireguard set up in all of our machines, and it is working propperly. 

The only way I could make our multicloud cluster work was disabling our Wireguard, and then everything works fine with:

**Server**:

`curl -sfL [https://get.k3s.io](https://get.k3s.io/) | INSTALL_K3S_EXEC="server --node-external-ip <SERVER_PUBLIC_IP> --flannel-backend=wireguard-native --flannel-external-ip" sh -s -`

**Agent**:

`curl -sfL [https://get.k3s.io](https://get.k3s.io/) | INSTALL_K3S_EXEC="agent --server https://<SERVER_PUBLIC_IP>:6443 --token <TOKEN> --node-external-ip <AGENT_PUBLICIP>" sh -s -`

With this, everything works properly, but I want to use our own Wireguard. When I configure K3s to use it, I can see that both nodes looks like ready, PODs are being deployed in both of them, but only the pods in the Server can interconnect between them. The PODs in the worker cannot connect to PODs in the worker, and PODs on Server can resolve PODs on worker, but they cannot stablish connection using hostnames.

If I do the deployment, and I enter in the pods and try to do manual connections with netcat, I connects succesfully with POD ip, Service IP, but not hostname, so I assume there is a DNS problem but I don’t know where:
```
root@pod-0:/node# nc -zv peer-34 5000
^C
root@pod-0:/node# nc -zv 10.42.1.27 5000
Connection to 10.42.1.27 5000 port [tcp/*] succeeded!
root@pod-0:/node# nc -zv 10.43.182.61 5000
Connection to 10.42.1.27 5000 port [tcp/*] succeeded!
root@pod-0:/node# nc -zv pod-34 5000
^C
root@pod-0:/node# nc -zv peer-34 5000
^C
```

The upper scenario is from a POD located in Server trying to connect a POD in Agent. The other way around doesn’t work. For example:
```
root@pod-1:/node# nc -zv peer-38 5000
^C
root@pod-1:/node# nc -zv 10.42.1.25 5000
^C
root@pod-1:/node# nc -zv 10.43.112.39 5000
^C
```

I am doing the following set up:

**Server**:

`curl -sfL [https://get.k3s.io](https://get.k3s.io/) | INSTALL_K3S_EXEC="--node-ip=<WIREGUARD_SERVER_IP> --advertise-address=<WIREGUARD_IP> --node-external-ip=<SERVER_PUBLIC_IP> --flannel-iface=wg0" sh -`

**Agent**:

`curl -sfL [https://get.k3s.io](https://get.k3s.io/) | K3S_URL=https://<WIREGUARD_SERVER_IP>:6443 K3S_TOKEN=<TOKEN> INSTALL_K3S_EXEC="--node-ip=<WIREGUARD_AGENT_IP> --node-external-ip=<AGENT_PUBLIC_IP> --flannel-iface=wg0" sh -`

I tried more configurations, but this looked the “most correct one” to me, and I didn’t want to flood this issue with every single configuration I tried. What I want to tell K3s is that it has to use our Wireguard interface, but for sure there is something wrong here.

**Information that I have read**:

This was very similar to my issue, but didn’t solve anything: https://github.com/alekc-go/flannel-fixer

This issue, that refers to the previous Issue: https://github.com/k3s-io/k3s/issues/1824

Regarding using `wireguard` (`wireguard-native`) as flannel backend, which makes K3s to crash because the wireguard interface is already set up: https://github.com/k3s-io/k3s/issues/1608

I assume when you are running `--flannel-backend=wireguard-native` it creates its own interface, which is when I am running into the following error: 

`level=fatal msg="flannel exited: failed to set up the route: failed to set interface flannel-wg to UP state: address already in use”`

Similar situation:

https://publish.reddit.com/embed?url=https://www.reddit.com/r/selfhosted/comments/mu6et4/has_anyone_setup_k3s_over_wireguard_is_it_possible/ih30pqj/

More issues/PR:

https://github.com/k3s-io/k3s/issues/2689

https://github.com/k3s-io/k3s/pull/6180

https://github.com/k3s-io/k3s/discussions/7384

https://hackmd.io/@WesleyCh3n/ntu_iot_k3s

**Checking `kubectl describe nodes`**:

Server:

```
…
Annotations:        
[`alpha.kubernetes.io/provided-node-ip:](http://alpha.kubernetes.io/provided-node-ip:) <WIREGUARD_SERVER-IP>
[flannel.alpha.coreos.com/backend-data:](http://flannel.alpha.coreos.com/backend-data:) {"VNI":1,"VtepMAC":"fe:34:5f:b9:9e:42"}
[flannel.alpha.coreos.com/backend-type:](http://flannel.alpha.coreos.com/backend-type:) vxlan
[flannel.alpha.coreos.com/kube-subnet-manager:](http://flannel.alpha.coreos.com/kube-subnet-manager:) true
[flannel.alpha.coreos.com/public-ip:](http://flannel.alpha.coreos.com/public-ip:) <WIREGUARD_SERVER-IP>
[k3s.io/external-ip:](http://k3s.io/external-ip:) <SERVER_PUBLIC_IP>
[k3s.io/hostname:](http://k3s.io/hostname:) <HOSTNAME>
[k3s.io/internal-ip:](http://k3s.io/internal-ip:) <WIREGUARD_SERVER-IP>
[k3s.io/node-args:](http://k3s.io/node-args:)
["server","--node-ip",<WIREGUARD_SERVER-IP>,"--advertise-address",<WIREGUARD_SERVER-IP>,"--node-external-ip",<SERVER_PUBLIC_IP>,"--flannel-iface","wg0"]
…
Addresses:
InternalIP:  <WIREGUARD_SERVER-IP>
ExternalIP:  <SERVER_PUBLIC_IP>
…
PodCIDR:                      10.42.0.0/24
PodCIDRs:                     10.42.0.0/24
```

In Agent information looks ok aswell.

**Kubernetes DNS resolution**:

https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/

```
$ sudo kubectl exec -i -t dnsutils -- nslookup kubernetes.default
Server:         10.43.0.10
Address:        10.43.0.10#53
Name:   kubernetes.default.svc.cluster.local
Address: 10.43.0.1
```

```
$ sudo kubectl exec -ti dnsutils -- cat /etc/resolv.conf
search default.svc.cluster.local svc.cluster.local cluster.local [statusim.net](http://statusim.net/)
nameserver 10.43.0.10
options ndots:5
```

DNS pod is running:

```$ sudo kubectl get pods --namespace=kube-system -l k8s-app=kube-dns```

```
$ sudo kubectl logs --namespace=kube-system -l k8s-app=kube-dns
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.override
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.server
```

```
$ kubectl get endpoints kube-dns --namespace=kube-system

NAME       ENDPOINTS                                  AGE
kube-dns   10.42.0.4:53,10.42.0.4:53,10.42.0.4:9153   10m
```

Adding log to CoreDNS:

```
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.override
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.server
[INFO] plugin/reload: Running configuration SHA512 = 0b4864bedbec2be9d5f56a213b8a6b63704dfe9ae1a318dda30b5aae0390e6943ed5526c895c41d120a3e93ad0c6302ce165e0d703b7df918cb7797453397d1f
[INFO] Reloading complete
[INFO] 127.0.0.1:51812 - 21722 "HINFO IN 7560458663601928980.4899293149017499055. udp 57 false 512" NXDOMAIN qr,rd,ra 132 0.000860068s
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.override
[WARNING] No files matching import glob pattern: /etc/coredns/custom/*.server
```

**Known Issues (From documentation)**:

*Some Linux distributions (e.g. Ubuntu) use a local DNS resolver by default (systemd-resolved). Systemd-resolved moves and replaces /etc/resolv.conf with a stub file that can cause a fatal forwarding loop when resolving names in upstream servers. This can be fixed manually by using kubelet's --resolv-conf flag to point to the correct resolv.conf (With systemd-resolved, this is /run/systemd/resolve/resolv.conf). kubeadm automatically detects systemd-resolved, and adjusts the kubelet flags accordingly.*

Checking `kubectl -n kube-system edit configmap coredns`:

`forward . /etc/resolv.conf`

It’s content:

```
nameserver 127.0.0.53
options edns0 trust-ad
search [statusim.net](http://statusim.net/)
```

Checking `/run/systemd/resolve/resolv.conf` :

```
nameserver 185.12.64.2
nameserver 185.12.64.1
nameserver 2a01:4ff:ff00::add:1
# Too many DNS servers configured, the following entries may be ignored.
nameserver 2a01:4ff:ff00::add:2
search [statusim.net](http://statusim.net/)
```

I tried modifying `/run/systemd/resolve/resolv.conf` to delete the two ipv6 addresses, and use it as an argument with *`--resolv-conf`*, but it is being overwriten for some reason I don’t understand .

**My questions are**:

- Am I doing something wrong in the setup?
- What is the point of using `--node-external-ip` in a situation like this? I guess it is useful for OpenLens or things like that? Or it is not related?
- Looks like it is a DNS problem, but I am not sure 100%
- If I have my own Wireguard, is it correct that I am using `vxlan` backend? I assume this is correctly abstracted?

**Expected behavior:**
Same behaviour as the set up with `--flannel-backend=wireguard-native` and our Wireguard deactivated.

**Actual behavior:**
Explained above.

**Additional context / logs:**
Above.

**Comment by brandond** [comment]
> I assume when you are running --flannel-backend=wireguard-native it creates its own interface, which is when I am running into the following error:
> 
> level=fatal msg="flannel exited: failed to set up the route: failed to set interface flannel-wg to UP state: address already in use”

yes, if you use the wireguard flannel backend, then flannel is assumed to own the wireguard configuration and interface management. If you want to configure wireguard on your own, then you should make sure that all the basics work - like the ability to resolve hostnames and pass traffic over the tunnels. You may want to experiment with different flannel backends (host-gw instead of vxlan, for example) and/or adjust the flannel interface MTU to get traffic to pass successfully through your tunnel.

**Reply by AlbertoSoutullo** [reply]
Hi @brandond and thank you for your answer.

I can confirm that our wireguard is working properly, in fact it is the network where our infra team gather metrics to monitorize all our machines (unless I am missing something here, will double check tomorrow).

Then, regarding `host-gw`, as far as I know this is not something that I could use, since in the documentation it states that `Requires direct layer 2 connectivity between all nodes in the cluster.`, and this is not the situation I am currently working on. With this I want to say that I am very confident that our Wireguard is set up properly and we have traffic through it.

Do you have any other idea I could check? 

Regards, 
Alberto

**Reply by brandond** [reply]
> I want to say that I am very confident that our Wireguard is set up properly and we have traffic through it.

You also said:
> With this, everything works properly, but I want to use our own Wireguard. When I configure K3s to use it, I can see that both nodes looks like ready, PODs are being deployed in both of them, but only the pods in the Server can interconnect between them. The PODs in the worker cannot connect to PODs in the worker, and PODs on Server can resolve PODs on worker, but they cannot stablish connection using hostnames.

This suggests that you don't have working DNS between your hosts? Are you relying on hosts files, or something else that isn't going to work within containers?

@manuelbuil may have additional suggestions.

**Comment by manuelbuil** [comment]
> What is the point of using --node-external-ip in a situation like this? I guess it is useful for OpenLens or things like that? Or it is not related?

You use that flag if there is an IP which is external and can't be seen from inside the node (e.g. in public clouds when you attach a public IP to the VM). If you have baremetal nodes and server/agent can communicate using the wireguard IP, then you don't need that flag.

> If I have my own Wireguard, is it correct that I am using vxlan backend? I assume this is correctly abstracted?

Yes it should work. It is no the most performing solution since you are doing double encapsulation. And you should be careful with MTUs, especially when doing IPv6, packets might get dropped but this is not your problem right now.

Your biggest problem is that in the agent, your pods can't do intracommunication. That is super weird because flannel creates a bridge and attaches the veth interfaces of the pods to it, so they should have direct communication. Can you verify that the bridge is created: `brctl show`

**Comment by AlbertoSoutullo** [comment]
> Your biggest problem is that in the agent, your pods can't do intracommunication. That is super weird because flannel creates a bridge and attaches the veth interfaces of the pods to it, so they should have direct communication. Can you verify that the bridge is created: `brctl show`

I did a fresh installation without the `--node-external-ip`. Checking `brctl show` gives me on Server:
```
bridge name     bridge id               STP enabled     interfaces
cni0            8000.52ccd048eb79       no              veth056d84bb
                                                        veth0f329837
                                                        veth138f5c1e
                                                        veth2947c2cd
                                                        veth3d807cfd
                                                        veth4eddcbbc
                                                        veth4f1692dc
                                                        veth5c9f7469
                                                        veth5cd82186
                                                        veth62e2cd76
                                                        veth6607c184
                                                        veth695b4166
                                                        veth6ddac805
                                                        veth6e8070c0
                                                        veth839c90dc
                                                        veth83ace3ef
                                                        veth91b0db74
                                                        veth9c9886ee
                                                        vethb2510e32
                                                        vethb2e7fc9a
                                                        vethb3f5ab04
                                                        vethc74f9d56
                                                        vethd3931b55
                                                        vethf7ee021c
docker0         8000.0242052da078       no
```
and on Agent:
```
bridge name     bridge id               STP enabled     interfaces
cni0            8000.7aa030e52641       no              veth01ce0b5d
                                                        veth0aabb677
                                                        veth137cb995
                                                        veth15d2c2bf
                                                        veth1acac883
                                                        veth3ebb17af
                                                        veth529238ba
                                                        veth5804fb82
                                                        veth5a2cf611
                                                        veth67e94963
                                                        veth79dd6a84
                                                        veth8afdec77
                                                        veth92cf42d0
                                                        veth96be4b8b
                                                        veth994fab34
                                                        vethcc5c5fa5
                                                        vethd284f0e3
                                                        vethdbb470ca
                                                        vethdbe20704
                                                        vethf03de694
                                                        vethfb62484e
                                                        vethfb9cdd94
docker0         8000.0242c81a1919       no
```
Which makes sense since I am deploying 40 PODs with one service each. So I assume these veths are for the services, and 4 extra for `kubernetes`, `kube-dns`, `metrics-server` and `traefik`, making the 44 I wrote.

**Reply by manuelbuil** [reply]
bridge cni0 will switch packets between veth interfaces, so you should be able to ping between pods, you can't?

**Reply by AlbertoSoutullo** [reply]
I can, but only with IP addresses. As I wrote in the first post, I am facing this behaviour:
```
root@pod-0:/node# nc -zv peer-34 5000
^C
root@pod-0:/node# nc -zv 10.42.1.27 5000
Connection to 10.42.1.27 5000 port [tcp/*] succeeded!
root@pod-0:/node# nc -zv 10.43.182.61 5000
Connection to 10.42.1.27 5000 port [tcp/*] succeeded!
root@pod-0:/node# nc -zv pod-34 5000
^C
root@pod-0:/node# nc -zv peer-34 5000
^C
```

```
root@pod-1:/node# nc -zv peer-38 5000
^C
root@pod-1:/node# nc -zv 10.42.1.25 5000
^C
root@pod-1:/node# nc -zv 10.43.112.39 5000
^C
```

Pod-0 is on Server, and peer-34 is on Agent. With IP address (either pod or service) I can reach to it, but not using the hostname. 
Pod-1 is on Agent, and here I can't reach any other pod, it doesn't matter if it is on agent or server.

Also, just to double check...

In the documentation is stated that I should have port 6443 opened on Servers, and 8472 for Flannel VXLAN. But it also says that we should have 51820 for Flannel Wireguard with IPv4.
As I am using VXLAN on top of our Wireguard, am I right asuming that I don't need 51820? Because as I read it, I am understanding that this is only if you are using the wireguard backend for Flannel. Just wanted to confirm that.

**Reply by manuelbuil** [reply]
I am now confused. You say: `I can, but only with IP addresses` but then you say `Pod-1 is on Agent, and here I can't reach any other pod, it doesn't matter if it is on agent or server`.

Parking for now hostname resolution and services and staying only with pod IPs, let's clarify this next 4 use cases:

* can podServer1 ping podAgent1 using IPs? I understand YES
* can podServer1 ping podServer2 using IPs? I understand YES
* can podAgent1 ping podAgent2 using IPs? Not sure
* can podAgent1 ping podServer1 using IPs? Not sure

**Reply by AlbertoSoutullo** [reply]
Sorry, I tried so many different things that I have pure chaos in my head.
Let me summarize from the fresh installation I did now, to be consistent with my answers. I removed `--node-external-ip`, so:

Server:
`curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--node-ip=<WIREGUARD_SERVER_IP> --advertise-address=<WIREGUARD_SERVER_IP> --flannel-iface=wg0" sh -`
Agent:
`curl -sfL https://get.k3s.io | K3S_URL=https://<WIREGUARD_SERVER_IP>:6443 K3S_TOKEN=<TOKEN> INSTALL_K3S_EXEC="--node-ip=<WIREGUARD_AGENT_IP> --flannel-iface=wg0" sh -`

So, answering to your questions:
- Can podServer1 ping podAgent1 using IPs? No it can not.
- Can podServer1 ping podServer2 using IPs? Yes it can (even with service name).
- Can podAgent1 ping podAgent2 using IPs? Yes it can (but not with service name).
- Can podAgent1 ping podServer1 using IPs? No it can not.

`brctl show` shows the interfaces in cni0 as in my previous comment.

**Reply by manuelbuil** [reply]
Ok. So, packets between nodes are getting dropped or misrouted for some reason. Assuming your network connectivity using your wireguard infrastructure works, if we think about the use case `podServer1 ping podAgent1`, the path should be:
1 - `echo request` packet leaves veth of podServer1 and reaches cni0
2 - `echo request` packet is routed to flannel.1 interface
3 - flannel.1 interface encaps the packet with vxlan and forwards it to wg0 interface
4 - wg0 interface encaps packet and sends it to wg0 on agent
5,6,7,8 - the same but on the other direction (e.g. wg0 on agent decaps and sends it to flannel.1-agent interface)

If I were you I would `tcpdump -ni $IF icmp -v` on the different involved interfaces: cni0, flannel.1, wg0, wg0-agent... and find out where the packet does not appear anymore. Note that in wg0 interface the packet is already encapsulated with vxlan, so you can't see icmp that easily, in that case filter on the vxlan port and the `-v` flag should output the icmp packet in the second line

**Comment by AlbertoSoutullo** [comment]
Ok. So I have faced the following situations:

**1. From POD-10 to POD-12 (both in Server):**
```
cni0: 13:50:57.999873 IP (tos 0x0, ttl 64, id 57694, offset 0, flags [DF], proto ICMP (1), length 84)
    10.42.0.12 > 10.42.0.15: ICMP echo request, id 11422, seq 1, length 64

cni0: 13:50:57.999948 IP (tos 0x0, ttl 64, id 48706, offset 0, flags [none], proto ICMP (1), length 84)
    10.42.0.15 > 10.42.0.12: ICMP echo reply, id 11422, seq 1, length 64
```
Looks like ok.

____________

**2. From POD-10 to POD-12 with service name (both in Server):**
```
cni0: 14:18:47.025976 IP (tos 0x0, ttl 64, id 63906, offset 0, flags [DF], proto ICMP (1), length 84)
    10.42.0.12 > 10.43.97.134: ICMP echo request, id 11471, seq 1, length 64
```
Command:
```
$ ping -c 1 peer-12
PING peer-12.my-namespace.svc.cluster.local (10.43.97.134) 56(84) bytes of data.
```

Looks like it doesnt reach `flannel.1`

____________

**3. From POD 10 to POD 0 (server -> node):**
```
cni0: 14:22:24.059942 IP (tos 0x0, ttl 64, id 22815, offset 0, flags [DF], proto ICMP (1), length 84)
    10.42.0.12 > 10.42.1.3: ICMP echo request, id 11477, seq 1, length 64

flannel.1: 14:22:24.060009 IP (tos 0x0, ttl 63, id 22815, offset 0, flags [DF], proto ICMP (1), length 84)
    10.42.0.12 > 10.42.1.3: ICMP echo request, id 11477, seq 1, length 64

wg0: ---
```
Command:
```
$ ping -c 1 10.42.1.3
PING 10.42.1.3 (10.42.1.3) 56(84) bytes of data.

--- 10.42.1.3 ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms
```


____________

**4. From POD-10 to POD-0 (with service name):**
```
cni0: 14:25:03.018005 IP (tos 0xc0, ttl 64, id 53795, offset 0, flags [none], proto ICMP (1), length 166)
    10.42.0.12 > 10.42.0.6: ICMP 10.42.0.12 udp port 42353 unreachable, length 146
        IP (tos 0x0, ttl 64, id 65379, offset 0, flags [DF], proto UDP (17), length 138)
    10.42.0.6.53 > 10.42.0.12.42353: 24271*- 1/0/0 peer-0.10k-namespace.svc.cluster.local. A 10.43.126.240 (110)
cni0: 14:25:03.018038 IP (tos 0xc0, ttl 63, id 53795, offset 0, flags [none], proto ICMP (1), length 166)
    10.42.0.12 > 10.42.0.6: ICMP 10.42.0.12 udp port 42353 unreachable, length 146
        IP (tos 0x0, ttl 64, id 65379, offset 0, flags [DF], proto UDP (17), length 138)
    10.42.0.6.53 > 10.42.0.12.42353: 24271*- 1/0/0 peer-0.10k-namespace.svc.cluster.local. A 10.43.126.240 (110)
cni0: 14:25:03.018078 IP (tos 0x0, ttl 64, id 14937, offset 0, flags [DF], proto ICMP (1), length 84)
    10.42.0.12 > 10.43.126.240: ICMP echo request, id 11481, seq 1, length 64
```
Command:
```
$ ping -c 1 peer-0
PING peer-0.my-namespace.svc.cluster.local (10.43.126.240) 56(84) bytes of data.

--- peer-0.my-namespace.svc.cluster.local ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms
```

So as a summary, looks like when I use IPs directly, the packet gets lost between `flannel.1` and `wg0`...
Then using service names, the packet stays on `cni0`, and goes directly to `CoreDNS`, and it goes back to the POD without going through any other interface, with some unreachable ports which is confusing me.

Edit: The information was checked with `tcpdump -ni cni0 icmp -v`, `tcpdump -ni flannel.1 icmp -v` and `tcpdump -ni wg0 icmp -v` commands.

**Comment by manuelbuil** [comment]
If pods from the agent can't reach pods from the server, name resolution will never work because your dns server is a pod running in the server. So, for the moment, I'd forget about that use case.

In the `From POD 10 to POD 0 (server -> node):`, could you sniff `wg0` interface without the `icmp` filter? You will not see any icmp packet because it is encapsulated using vxlan headers. For example: `tcpdump -ni wg0 port 8472 -v`

**Reply by AlbertoSoutullo** [reply]
Ok I checked a couple of times, and looks like the packet does all steps 1,2,3,4 and reaches 5. I can see it reaches `wg0` in the agent. But unless I have to check it elsewhere or with any other command, the packet is missing between 5 and 6, so between `wg0` and `flannel.1` in the agent.

Just to double check, I am using the same commands in both machines, so in the agent I am checking the packet with `tcpdump -ni flannel.1 icmp -v`, and it is not displaying anyting. I guess it should display as `flannel.1` is in charge of de-encapsulate the packet?

Edit: I am very grateful for this guidance btw.

**Reply by manuelbuil** [reply]
Yep, you should see it there. At least we know that nothing in between the nodes is dropping it (e.g. because of wrong MTU). My first guess would be iptables dropping those packets or the forwarding not enabled in the agent `/proc/sys/net/ipv4/ip_forward`

**Reply by AlbertoSoutullo** [reply]
Ok, it is related to iptables for sure, it has duplicated lines and it is quite big. I have been reading the documentation if [k3s with iptables](https://docs.k3s.io/known-issues#iptables) and it says that:
- K3s recommend using newer iptables (such as 1.6.1+) to avoid issues.
- Versions 1.8.0-1.8.4 have known issues.

I am using **iptables v1.8.7 (nf_tables)** on both machines. 

Also as extra info:
K3s Version:
k3s version v1.27.6+k3s1 (https://github.com/k3s-io/k3s/commit/bd04941a294793ec92e8703d5e5da14107902e88)
go version go1.20.8

Node(s) CPU architecture, OS, and Version:
Server: Linux 5.15.0-75-generic https://github.com/k3s-io/k3s/issues/82-Ubuntu SMP Tue Jun 6 23:10:23 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
Agent: Linux 5.15.0-84-generic https://github.com/k3s-io/k3s/issues/93-Ubuntu SMP Tue Sep 5 17:16:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

I think I will reinstall K3s with the flag that you mention [here](https://docs.k3s.io/advanced#old-iptables-versions). What I don't really understand is how this is managed by K3s.
So I will have my own iptables, and if I use `--prefer-bundled-bin`, it will use it's own iptables? Woulnd't they have conflicts?

**Reply by manuelbuil** [reply]
I think it would be best if you could identify what iptables rule is dropping your packets and then try to figure out why you have that rule there. Could it be that firewalld is enabled?

**Reply by AlbertoSoutullo** [reply]
Is there any documentation to check which rules from K3s are added to Iptables? Because there are a lot, and I am not 100% sure which ones are from K3s.
Edit: Looks to me that there is one missing rule that allows packets from port 8472 in the VPN chain, but for some reason each time I am adding that rule it's get deleted. I have tried uninstalling K3s, adding the rule and reinstalling it, but for some reason the rule gets deleted after the installation. It is bugging me a lot. Will share here once I have more info...

**Comment by manuelbuil** [comment]
When adding the rule that allows packets from port 8472 things work?

**Reply by AlbertoSoutullo** [reply]
Well, looks like each time I install K3s, the rule is deleted from the chain VPN is also removed.
Before:
```
Chain VPN (2 references)
 pkts bytes target     prot opt in     out     source               destination
17397 1044K ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:8001 match-set metrics.hq src /* Netdata Metrics */
  250 15000 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:8002 match-set metrics.hq src /* MTR Metrics */
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:6443 match-set vacdst.misc src /* Kubernetes Supervisor */
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:10250 match-set vacdst.misc src /* Kubernetes Metrics */
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:8472 match-set vacdst.misc src /* Kubernetes Flannel VXLAN */
```
After: 
```
Chain VPN (2 references)
 pkts bytes target     prot opt in     out     source               destination
    7   420 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:8001 match-set metrics.hq src /* Netdata Metrics */
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:8002 match-set metrics.hq src /* MTR Metrics */
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:6443 match-set vacdst.misc src /* Kubernetes Supervisor */
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:10250 match-set vacdst.misc src /* Kubernetes Metrics */
```

Also, apart from all the new rules and tables that K3s is adding, I am doing a `| grepp 8472` to see if maybe the rule was moved, but it just disappeared. Is this a normal behaviour?

Edit: It happened on both machines that I am using. Before Installing K3s, the rule was there. After the instalation, it is not.

Edit2: Just realized 8472 is udp and not tcp, but I don't think this is the issue of why it is getting removed right?

Edit3: We have changed the protocol to udp, and the rule disapeared when running the uninstall script. So looks like installing/uninstalling K3s is modifiying that rule.

**Comment by AlbertoSoutullo** [comment, marked as answer]
I come with some good news. I managet to get it working. The root of the problem was the script `/usr/local/bin/k3s-killall.sh` which is called when you call the main uninstall scipt, either in Server or Agent.
In that script I found this:
`iptables-save | grep -v KUBE- | grep -v CNI- | grep -iv flannel | iptables-restore`

That is why our rule in iptables was getting deleted on each install/uninstall, as it has flannel in the comment.
`0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:8472 match-set vacdst.misc src /* Kubernetes Flannel VXLAN */`

After checking that the rule was there when the cluster is initiated, everything worked perfectly. I will change the comment in the rule but it was quite a headache to discover what was happening...

**Reply by manuelbuil** [reply]
awesome!

**Reply by AlbertoSoutullo** [reply]
Imo this kind of behaviour should be documented because it is very hard to find this. Or if it is documented, I didn't find it anywhere.

Still, thank you very much for your guidance.

Pd: Oscense power.
